ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java

```diff
@@ -458,8 +445,25 @@ String getInternalColumnForExpression(final Expression expression) {
       return internalNameToIndexMap;
     }
 
-    private Map<String, String> getExpressionToInternalColumnNameMap() {
-      return expressionToInternalColumnNameMap;
+    private Expression resolveToInternal(final Expression exp) {
+      if (exp instanceof FunctionCall) {
+        final FunctionCall funcCall = (FunctionCall) exp;
+        final List<Expression> internalArgs = getInternalExpressionList(funcCall.getArguments());
+        return new FunctionCall(
+            funcCall.getLocation(), funcCall.getName(), funcCall.getWindow(), funcCall.isDistinct(),
+            internalArgs);
+      }
+
+      final String name = expressionToInternalColumnNameMap.get(exp.toString());
+      if (name != null) {
+        return new QualifiedNameReference(QualifiedName.of(name));
+      }
+
+      if (exp instanceof DereferenceExpression) {
+        throw new KsqlException("GROUP BY expression must be part of SELECT: " + exp.toString());
```

Comment:
Few questions/comments here:
    - Is this error message true? We've made changes in this PR to support exactly this and have tests to verify. Would be more accurate to say "GROUP BY expression not found amongst internal columns"
    - This seems like an internal error to me. What statement could a user issue that would produce this?

Reply:
I'll look into it. I think you're right - its more an internal error, but I need to refresh my memory of this. I'll address in the follow up PR.

Reply:
So the latest has changed this code to a visitor pattern and still throws an exception, even though its not currently possible to hit this point in the code.  This specific exception is still better than an NPE if/when code changes in the future.